### PR TITLE
statistics: not always COW to update stats cache

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/45367

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/handle.go [hunk 1]
```diff
@@ -1007,12 +1007,21 @@ func (h *Handle) statsCacheLen() int {
...
+}
+
+// copyAndUpdateStatsCache overrides the global statsCache with a new one, it may fail
 // if the global statsCache has been modified by others already.
 // Callers should add retry loop if necessary.
-func (h *Handle) updateStatsCache(newCache *cache.StatsCache) (updated bool) {
-	h.statsCache.Replace(newCache)
-	handle_metrics.CostGauge.Set(float64(newCache.Cost()))
+func (h *Handle) copyAndUpdateStatsCache(newCache *cache.StatsCache, tables []*statistics.Table, deletedIDs []int64, opts ...cache.TableStatsOpt) (updated bool) {
+	if enableQuota := config.GetGlobalConfig().Performance.EnableStatsCacheMemQuota; enableQuota {
```
Comment (qw4990) [inline statistics/handle/handle.go:1019]: Better to add some comments for this .

File: statistics/handle/handle.go [hunk 2]
```diff
@@ -1007,12 +1007,21 @@ func (h *Handle) statsCacheLen() int {
...
+	handle_metrics.CostGauge.Set(float64(h.statsCache.Load().Cost()))
+}
+
+// copyAndUpdateStatsCache overrides the global statsCache with a new one, it may fail
 // if the global statsCache has been modified by others already.
 // Callers should add retry loop if necessary.
-func (h *Handle) updateStatsCache(newCache *cache.StatsCache) (updated bool) {
-	h.statsCache.Replace(newCache)
-	handle_metrics.CostGauge.Set(float64(newCache.Cost()))
+func (h *Handle) copyAndUpdateStatsCache(newCache *cache.StatsCache, tables []*statistics.Table, deletedIDs []int64, opts ...cache.TableStatsOpt) (updated bool) {
```
Comment (qw4990) [inline statistics/handle/handle.go:1018]: How about using `updateStatsCache` since this method doesn't always copy this cache...?

**Review by qw4990 — APPROVED** [review]

File: statistics/handle/handle.go (hunk 2 above)
Comment (hawkingrei) [reply statistics/handle/handle.go]: Done

File: statistics/handle/handle.go (hunk 1 above)
Comment (hawkingrei) [reply statistics/handle/handle.go:1020]: Done

File: statistics/handle/cache/statscacheinner.go [hunk 3]
```diff
@@ -172,6 +172,32 @@ func (sc *StatsCache) CopyAndUpdate(tables []*statistics.Table, deletedIDs []int
...
+		id := tbl.PhysicalID
+		if option.byQuery {
+			sc.c.Put(id, tbl, true)
+		} else {
+			sc.c.Put(id, tbl, false)
+		}
+	}
+	for _, id := range deletedIDs {
+		sc.c.Del(id)
+	}
```
Comment (time-and-fate) [inline statistics/handle/cache/statscacheinner.go:191]: Seems we are treating `StatsCache` as thread-safe here. Is this expected?

**Review by time-and-fate — APPROVED** [review]
Others LGTM. 
`(sc *StatsCache) Update` is safe for now, but we should make the `MapCache` thread-safe in the future.

File: statistics/handle/cache/statscacheinner.go (hunk 3 above)
Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:191]: Yes, mapCache will be the thread-safe in the plan.

**Pull request merged**